### PR TITLE
graph: add type to allow defined conversion directed -> undirected

### PR DESCRIPTION
**Pull request by kortschak**
There are a couple of places this could go; either simple or graph. It seems to me that it's fundamental enough that it should be in graph.

@vladimir-ch Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
The converse type is not required since it is trivially implemented for most cases as:

```
type directed struct{ graph.Undirected }

func (g directed) To(n graph.Node) []graph.Node       { return g.From(n) }
func (g directed) HasEdgeFromTo(u, v graph.Node) bool { return g.HasEdgeBetween(u, v) }
```

File: undirect.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	// weights in G are resolved. A merge
+	// is performed if at least one edge
+	// exists between the nodes being
+	// considered.
+	// The order of weight parameters
+	// passed to Merge is not defined, so
+	// the function should be commutative.
+	// If Merge is nil, the arithmetic
+	// mean is used to merge weights.
+	Merge func(x, y float64) float64
```
Comment (kortschak) [inline undirect.go]: An alternative approach here is to define it as `func(x float64, xe Edge, y float64, ye Edge) float64`. This allows the state of the edges to be passed out of band, and for the user to make subtler decisions about what the result should be. What do you think?

File: undirect.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+// license that can be found in the LICENSE file.
+
+package graph
+
+// Undirect converts a directed graph to an undirected graph, resolving
+// edge weight conflicts.
+type Undirect struct {
+	G Directed
+
+	// Absent if the value used to
```
Comment (vladimir-ch) [inline undirect.go]: `is`

Comment (kortschak) [inline undirect.go]: Done.

Comment (vladimir-ch) [inline undirect.go]: Not really :-)

Comment (kortschak) [inline undirect.go]: Try again.

File: undirect.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	return g.Merge(f, r), ok
+}
+
+// EdgePair is an opposed pair of directed edges.
+type EdgePair struct {
+	E [2]Edge
+	W float64
+}
+
+// From returns the from node of the first non-nil edge, nil.
```
Comment (vladimir-ch) [inline undirect.go]: ... edge, _or_ nil.

Comment (kortschak) [inline undirect.go]: Done, with trepidation.

File: undirect.go (hunk 1 above)
Comment (vladimir-ch) [inline undirect.go]: I think it is fine to pass the edges, they can be easily ignored if not necessary.

File: undirect.go (hunk 3 above)
Comment (vladimir-ch) [inline undirect.go]: By trepidation you refer to the minimalism of the wording?

**Comment by kortschak** [comment]
PTAL

I have used `(x, y float64, xe, ye graph.Edge)` rather than what I suggested above purely to reduce the signature length. Comments on wording around that field gratefully accepted.

File: undirect.go (hunk 3 above)
Comment (kortschak) [inline undirect.go]: No, the trepidation associated with an unreasonably high probability that I have made another mistake in trying to fix one.

**Comment by vladimir-ch** [comment]
> to reduce the signature length

Yes, much better.

> Comments on wording around that field gratefully accepted.

It think it is ok.

LGTM

**Pull request merged**